{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "47e61338-bafe-4e50-a9ec-cbfa9dc34b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "from graph_vis import display_graph\n",
    "from open_neo4j import open_driver, get_graph, open_gds, make_graph_projection\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.simplefilter('ignore')\n",
    "\n",
    "node_pattern = re.compile(r'\\(n[0-9]+:[A-Za-z]+\\)')\n",
    "edge_pattern = re.compile(r'\\[:[A-Za-z_]+\\]')\n",
    "\n",
    "driver = open_driver()\n",
    "spoke = open_gds(\"spoke\") # spoke\n",
    "genelab = open_gds(\"spoke-genelab-v0.0.2\") # genelab database\n",
    "composite = open_gds(\"compositenasa\") # connected composite db"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26cce78f-c712-4c2d-90f0-41b98403ae43",
   "metadata": {},
   "source": [
    "# MetaGraph for GeneLab DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d647db3b-51f9-47f6-94e5-04132c7a91c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5c85c7bbebfc4273bb05aee0f69a18fa",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GraphWidget(layout=Layout(height='500px', width='100%'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metagraph = get_graph(driver, 'call apoc.meta.graph', 'spoke-genelab-v0.0.2')\n",
    "display_graph(metagraph, hide_edge=False, meta=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74337074-2773-4f89-aae1-e31604109731",
   "metadata": {},
   "source": [
    "# Check node count in GeneLab "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5be2e72e-2c8b-4c20-8aae-11e059935e7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGene</td>\n",
       "      <td>105984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene</td>\n",
       "      <td>19755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assay</td>\n",
       "      <td>1698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Study</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Anatomy</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MetaNode</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CellType</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      label   count\n",
       "0     MGene  105984\n",
       "1      Gene   19755\n",
       "2     Assay    1698\n",
       "3     Study      69\n",
       "4   Anatomy      25\n",
       "5  MetaNode       6\n",
       "6  CellType       3"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (node)\n",
    "RETURN\n",
    "  head(labels(node)) AS label,\n",
    "  count(*) AS count\n",
    "ORDER BY count DESC\n",
    "\"\"\"\n",
    "\n",
    "df = genelab.run_cypher(query)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "294380e3-cb73-4be3-b5f4-5c46877a5590",
   "metadata": {},
   "source": [
    "# Check edge count in GeneLab "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "679ca588-4a5a-4637-b2f9-2e28dca95444",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rel_type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEASURED_AmMG</td>\n",
       "      <td>5046101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IS_ORTHOLOG_MGiG</td>\n",
       "      <td>60972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PERFORMED_SpA</td>\n",
       "      <td>1698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INVESTIGATED_AiA</td>\n",
       "      <td>1665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>INVESTIGATED_AiCT</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MetaRelationship</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            rel_type    count\n",
       "0      MEASURED_AmMG  5046101\n",
       "1   IS_ORTHOLOG_MGiG    60972\n",
       "2      PERFORMED_SpA     1698\n",
       "3   INVESTIGATED_AiA     1665\n",
       "4  INVESTIGATED_AiCT       55\n",
       "5   MetaRelationship        5"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH ()-[rel]->()\n",
    "RETURN\n",
    "  type(rel) AS rel_type,\n",
    "  count(*) AS count\n",
    "ORDER BY count DESC\n",
    "\"\"\"\n",
    "\n",
    "df = genelab.run_cypher(query)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fb3eecc-9dfa-48d6-b93f-a3e372183dfe",
   "metadata": {},
   "source": [
    "# Find the number of assays that are connected to Anatomy nodes in SPOKE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "eb581eaa-084f-4156-85ef-da003e96fd12",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Node_Name</th>\n",
       "      <th>n_assays</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UBERON:0000178</td>\n",
       "      <td>blood</td>\n",
       "      <td>335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UBERON:0000468</td>\n",
       "      <td>multicellular organism</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UBERON:0002107</td>\n",
       "      <td>liver</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UBERON:0002106</td>\n",
       "      <td>spleen</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UBERON:0004538</td>\n",
       "      <td>left kidney</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UBERON:0001115</td>\n",
       "      <td>left lobe of liver</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UBERON:0001377</td>\n",
       "      <td>quadriceps femoris</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UBERON:0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>UBERON:0001911</td>\n",
       "      <td>mammary gland</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UBERON:0001389</td>\n",
       "      <td>soleus muscle</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>UBERON:0002370</td>\n",
       "      <td>thymus</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>UBERON:0000033</td>\n",
       "      <td>head</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>brain</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>UBERON:0002371</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>UBERON:0001388</td>\n",
       "      <td>gastrocnemius</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>UBERON:0001158</td>\n",
       "      <td>descending colon</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>UBERON:0001386</td>\n",
       "      <td>extensor digitorum longus</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>UBERON:0008951</td>\n",
       "      <td>left lung lobe</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>UBERON:0000948</td>\n",
       "      <td>heart</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>UBERON:0000922</td>\n",
       "      <td>embryo</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>UBERON:0000970</td>\n",
       "      <td>eye</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>UBERON:0001379</td>\n",
       "      <td>vastus lateralis</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UBERON:0002369</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>UBERON:0001133</td>\n",
       "      <td>cardiac muscle tissue</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>UBERON:0004548</td>\n",
       "      <td>left eye</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Node                  Node_Name  n_assays\n",
       "0   UBERON:0000178                      blood       335\n",
       "1   UBERON:0000468     multicellular organism       291\n",
       "2   UBERON:0002107                      liver       232\n",
       "3   UBERON:0002106                     spleen       176\n",
       "4   UBERON:0004538                left kidney        97\n",
       "5   UBERON:0001115         left lobe of liver        91\n",
       "6   UBERON:0001377         quadriceps femoris        79\n",
       "7   UBERON:0000966                     retina        70\n",
       "8   UBERON:0001911              mammary gland        63\n",
       "9   UBERON:0001389              soleus muscle        36\n",
       "10  UBERON:0002370                     thymus        33\n",
       "11  UBERON:0000033                       head        28\n",
       "12  UBERON:0000955                      brain        26\n",
       "13  UBERON:0002371                bone marrow        22\n",
       "14  UBERON:0001388              gastrocnemius        17\n",
       "15  UBERON:0001158           descending colon        15\n",
       "16  UBERON:0001386  extensor digitorum longus        15\n",
       "17  UBERON:0008951             left lung lobe        15\n",
       "18  UBERON:0000948                      heart         7\n",
       "19  UBERON:0000922                     embryo         6\n",
       "20  UBERON:0000970                        eye         3\n",
       "21  UBERON:0001379           vastus lateralis         3\n",
       "22  UBERON:0002369              adrenal gland         3\n",
       "23  UBERON:0001133      cardiac muscle tissue         1\n",
       "24  UBERON:0004548                   left eye         1"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "CALL {\n",
    "    USE compositenasa.glds\n",
    "    MATCH (n0:Anatomy)-[]-(n1:Assay)\n",
    "    WITH n0, SIZE(COLLECT(DISTINCT n1)) as n_assays\n",
    "    RETURN COLLECT(n0.identifier) AS ids, n_assays\n",
    "}\n",
    "CALL {\n",
    "    USE compositenasa.gldsspoke\n",
    "    WITH ids\n",
    "    MATCH (n:Anatomy) WHERE n.identifier IN ids\n",
    "    RETURN DISTINCT n.identifier as Node, n.name as Node_Name\n",
    "}\n",
    "RETURN DISTINCT Node, Node_Name, n_assays ORDER BY n_assays DESC\n",
    "\"\"\"\n",
    "composite.run_cypher(query) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a25b79ef-da0b-4923-9ebf-0060cf8a4a38",
   "metadata": {},
   "source": [
    "# Forwarding GeneLab values to SPOKE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebb919b5-6107-4417-9df5-76d51b742ea1",
   "metadata": {},
   "source": [
    "#### Here we find the Assays that measured gene expression in the eye. Then we group the upregulated Genes by the Symptoms they are connected to in SPOKE. Finally, we calculate the average log2fc for each Symptom."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b4a6fd29-66b9-4fc2-a052-0d011aad070a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Node_Name</th>\n",
       "      <th>avg_log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D007022</td>\n",
       "      <td>Hypotension</td>\n",
       "      <td>3.532166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C537048</td>\n",
       "      <td>Renotubular dysgenesis</td>\n",
       "      <td>3.532166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D051436</td>\n",
       "      <td>Chronic kidney disease</td>\n",
       "      <td>2.145078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D053099</td>\n",
       "      <td>Azotemia</td>\n",
       "      <td>1.910546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D003318</td>\n",
       "      <td>Corneal opacity</td>\n",
       "      <td>1.795103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C535342</td>\n",
       "      <td>Zonular cataract</td>\n",
       "      <td>1.739770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>C566162</td>\n",
       "      <td>Aculeiform cataract</td>\n",
       "      <td>1.739770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>C563426</td>\n",
       "      <td>Pulverulent cataract</td>\n",
       "      <td>1.739770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>D004844</td>\n",
       "      <td>Epistaxis</td>\n",
       "      <td>1.631696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>D008595</td>\n",
       "      <td>Menorrhagia</td>\n",
       "      <td>1.631696</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Node               Node_Name  avg_log2fc\n",
       "0  D007022             Hypotension    3.532166\n",
       "1  C537048  Renotubular dysgenesis    3.532166\n",
       "2  D051436  Chronic kidney disease    2.145078\n",
       "3  D053099                Azotemia    1.910546\n",
       "4  D003318         Corneal opacity    1.795103\n",
       "5  C535342        Zonular cataract    1.739770\n",
       "6  C566162     Aculeiform cataract    1.739770\n",
       "7  C563426    Pulverulent cataract    1.739770\n",
       "8  D004844               Epistaxis    1.631696\n",
       "9  D008595             Menorrhagia    1.631696"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "CALL {\n",
    "    USE compositenasa.glds\n",
    "    MATCH (n0:Anatomy {identifier:'UBERON:0000970'})-[]-(n1:Assay)-[r:MEASURED_AmMG {direction:'%s'}]->(n2:MGene)-[]->(n3:Gene)\n",
    "    RETURN DISTINCT n3.identifier AS ids, AVG(r.log2fc) as mean_log2fc\n",
    "}\n",
    " \n",
    "CALL {\n",
    "    USE compositenasa.gldsspoke\n",
    "    WITH ids\n",
    "    MATCH (n0:Gene)-[]-(n1:Symptom) WHERE n0.identifier = ids\n",
    "    RETURN DISTINCT n1.identifier as Node, n1.name as Node_Name, n0\n",
    "}\n",
    "WITH Node, Node_Name, AVG(mean_log2fc) as avg_log2fc\n",
    "RETURN DISTINCT Node, Node_Name, avg_log2fc ORDER BY avg_log2fc DESC LIMIT 10 \n",
    "\"\"\"\n",
    "composite.run_cypher(query % 'upregulates')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff2ddab4-1a69-4233-9f67-617ebedb958d",
   "metadata": {},
   "source": [
    "# MetaGraph for GeneLab DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b615bbec-02d8-48d1-a1e6-a15b38539084",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find assay where factor 1 = space flight\n",
    "query = \"\"\"MATCH (n:Assay) WHERE \"['Space Flight']\" in n.factors_1 AND \"['Ground Control']\" in n.factors_2 RETURN DISTINCT n.identifier as n\"\"\"\n",
    "sf_assays = list(genelab.run_cypher(query).n.unique())\n",
    "\n",
    "# end disease\n",
    "disease = 'DOID:9428' # intracranial-hypertension\n",
    "\n",
    "# MetaPaths from GeneLab\n",
    "gs_metapath = \"(a:Assay)-[r:MEASURED_AmMG]->(:MGene)-[]->(n:Gene)\"\n",
    "gs_filter = \"\\nWHERE r.direction = 'upregulates' AND a.identifier IN %s\" % sf_assays"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c677094-985b-4b83-bf9d-6e4bc663620e",
   "metadata": {},
   "source": [
    "# DWPC\n",
    "##### Degree Weighted Path Count (DWPC) scores the connectivity between nodes by taking the number of forward and reverse degrees into consideration.\n",
    "##### Read about how DWPC was used to predict [Disease-Gene](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004259) and [Disease-Compound](https://elifesciences.org/articles/26726) relationships.\n",
    "![alt text](jupyter_example_images_dwpc.png \"DWPC\")\n",
    "### Look at weighted DWPC for BiologicalProcesses using upregulated genes in Space Flight v Control assays\n",
    "![alt text](bp_dwpc_metagraph.png \"BP DWPC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "20ca8ebf-7d18-453a-adda-0ea6cadbdcc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0fa9280dd01f459cb090e92176857e50",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GraphWidget(layout=Layout(height='500px', width='100%'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>weighted_DWPC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0019835</td>\n",
       "      <td>cytolysis</td>\n",
       "      <td>3.962651e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0032200</td>\n",
       "      <td>telomere organization</td>\n",
       "      <td>2.245425e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0060333</td>\n",
       "      <td>type II interferon-mediated signaling pathway</td>\n",
       "      <td>2.112674e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0048877</td>\n",
       "      <td>homeostasis of number of retina cells</td>\n",
       "      <td>2.007509e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0051095</td>\n",
       "      <td>regulation of helicase activity</td>\n",
       "      <td>1.941231e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0006954</td>\n",
       "      <td>inflammatory response</td>\n",
       "      <td>1.627973e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0071357</td>\n",
       "      <td>cellular response to type I interferon</td>\n",
       "      <td>1.504975e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0000122</td>\n",
       "      <td>negative regulation of transcription by RNA po...</td>\n",
       "      <td>1.482725e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:1903706</td>\n",
       "      <td>regulation of hemopoiesis</td>\n",
       "      <td>1.445508e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0034340</td>\n",
       "      <td>response to type I interferon</td>\n",
       "      <td>1.363305e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   identifier                                               name  \\\n",
       "0  GO:0019835                                          cytolysis   \n",
       "1  GO:0032200                              telomere organization   \n",
       "2  GO:0060333      type II interferon-mediated signaling pathway   \n",
       "3  GO:0048877              homeostasis of number of retina cells   \n",
       "4  GO:0051095                    regulation of helicase activity   \n",
       "5  GO:0006954                              inflammatory response   \n",
       "6  GO:0071357             cellular response to type I interferon   \n",
       "7  GO:0000122  negative regulation of transcription by RNA po...   \n",
       "8  GO:1903706                          regulation of hemopoiesis   \n",
       "9  GO:0034340                      response to type I interferon   \n",
       "\n",
       "   weighted_DWPC  \n",
       "0   3.962651e-07  \n",
       "1   2.245425e-07  \n",
       "2   2.112674e-07  \n",
       "3   2.007509e-07  \n",
       "4   1.941231e-07  \n",
       "5   1.627973e-07  \n",
       "6   1.504975e-07  \n",
       "7   1.482725e-07  \n",
       "8   1.445508e-07  \n",
       "9   1.363305e-07  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from spokecloud import get_inner_dwpc_degree_str, OVERALL_WEIGHTED_DWPC_QUERY , plot_dwpc_paths\n",
    "spoke_metapath = \"(n0:Gene)-[:PARTICIPATES_GpBP]-(n1:BiologicalProcess)-[:PARTICIPATES_GpBP]-(n2:Gene)-[:ASSOCIATES_DaG]-(n3:Disease {identifier:'%s'})\"\n",
    "spoke_metapath = spoke_metapath % disease\n",
    "weight_prop = 'log2fc'\n",
    "edges = [edge[2:-1] for edge in edge_pattern.findall(spoke_metapath)]\n",
    "degree_str = get_inner_dwpc_degree_str(edges)\n",
    "group_by_node = 'n1'\n",
    "\n",
    "query = OVERALL_WEIGHTED_DWPC_QUERY % (gs_metapath, gs_filter, spoke_metapath, gs_metapath[:-1], gs_filter, spoke_metapath, degree_str, group_by_node, group_by_node, weight_prop, group_by_node, group_by_node)\n",
    "dwpc_df = composite.run_cypher(query)\n",
    "inner_nodes = list(dwpc_df.identifier.values[:8])\n",
    "plot_dwpc_paths(genelab, driver, spoke_metapath, group_by_node, gs_metapath, gs_filter, inner_nodes, max_show = 10)\n",
    "dwpc_df.head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0266cc53-11c4-469a-9e73-8935f22ff6f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CALL {\n",
      "    USE compositenasa.glds\n",
      "    MATCH (eye:Anatomy)-[]-(a:Assay)-[r:MEASURED_AmMG]->(:MGene)-[]->(n:Gene) \n",
      "WHERE r.direction = 'upregulates' AND a.identifier IN ['GLDS-255-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-102-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-47-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-137-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-162-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-100-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-326-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-173-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-423-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-21-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-4-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-352-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-25-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-421-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-161-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-194-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-513-8e37d1fa27e648b5dd35d0c8f19d6408'] AND eye.identifier IN ['UBERON:0000970', 'UBERON:0004548', 'UBERON:0000966']\n",
      "    RETURN COLLECT(DISTINCT n.identifier) AS ids\n",
      "}\n",
      "CALL {\n",
      "    USE compositenasa.gldsspoke\n",
      "    WITH ids\n",
      "    MATCH (n0:Gene)-[:PARTICIPATES_GpBP]-(n1:BiologicalProcess)-[:PARTICIPATES_GpBP]-(n2:Gene)-[:ASSOCIATES_DaG]-(n3:Disease {identifier:'DOID:9428'}) WHERE n0.identifier IN ids\n",
      "    RETURN DISTINCT n0.identifier AS overlap_nodes\n",
      "}\n",
      "CALL {\n",
      "    WITH overlap_nodes\n",
      "    USE compositenasa.glds\n",
      "    MATCH path_1=(eye:Anatomy)-[]-(a:Assay)-[r:MEASURED_AmMG]->(:MGene)-[]->(n:Gene {identifier:overlap_nodes})\n",
      "WHERE r.direction = 'upregulates' AND a.identifier IN ['GLDS-255-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-102-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-47-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-137-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-162-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-100-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-326-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-173-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-423-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-21-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-4-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-352-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-25-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-421-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-161-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-194-8e37d1fa27e648b5dd35d0c8f19d6408', 'GLDS-513-8e37d1fa27e648b5dd35d0c8f19d6408'] AND eye.identifier IN ['UBERON:0000970', 'UBERON:0004548', 'UBERON:0000966']\n",
      "    RETURN  n { .* } AS p, r\n",
      "}\n",
      "CALL {\n",
      "    WITH overlap_nodes\n",
      "    USE compositenasa.gldsspoke\n",
      "    WITH COLLECT(overlap_nodes) as overlap\n",
      "    MATCH path_2=(n0:Gene)-[:PARTICIPATES_GpBP]-(n1:BiologicalProcess)-[:PARTICIPATES_GpBP]-(n2:Gene)-[:ASSOCIATES_DaG]-(n3:Disease {identifier:'DOID:9428'}) WHERE n0.identifier IN overlap\n",
      "    WITH [size([(n0)-[:PARTICIPATES_GpBP]-() | n0]), size([()-[:PARTICIPATES_GpBP]-(n1) | n1]),\n",
      "size([(n1)-[:PARTICIPATES_GpBP]-() | n1]), size([()-[:PARTICIPATES_GpBP]-(n2) | n2]),\n",
      "size([(n2)-[:ASSOCIATES_DaG]-() | n2]), size([()-[:ASSOCIATES_DaG]-(n3) | n3])] AS degrees, n0, n1\n",
      "    WITH n0, n1, sum(reduce(pdp = 1.0, d in degrees| pdp * d ^ -0.7)) AS DWPC\n",
      "    RETURN n0, n1, DWPC\n",
      "}\n",
      "WITH n1, sum(r.log2fc*DWPC) as weighted_DWPC\n",
      "RETURN DISTINCT n1.identifier as identifier, n1.name as name , weighted_DWPC ORDER BY weighted_DWPC DESC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(query)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3349e00c-de2b-456b-ad9e-a9377e39f474",
   "metadata": {},
   "source": [
    "### Look at weighted DWPC for Symptoms using upregulated genes in Space Flight v Control assays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4e6c1fc7-09cd-4028-b183-0a5ebd321807",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a1350d01fbd54ef98db8b4ef3abde20e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GraphWidget(layout=Layout(height='500px', width='100%'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>weighted_DWPC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D003128</td>\n",
       "      <td>Coma</td>\n",
       "      <td>0.000651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D014474</td>\n",
       "      <td>Unconsciousness</td>\n",
       "      <td>0.000410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D006935</td>\n",
       "      <td>Hypercapnia</td>\n",
       "      <td>0.000353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D006261</td>\n",
       "      <td>Headache</td>\n",
       "      <td>0.000277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D006985</td>\n",
       "      <td>Hyperventilation</td>\n",
       "      <td>0.000257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>D012021</td>\n",
       "      <td>Reflex, Abnormal</td>\n",
       "      <td>0.000256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>D015354</td>\n",
       "      <td>Vision, Low</td>\n",
       "      <td>0.000185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>D004172</td>\n",
       "      <td>Diplopia</td>\n",
       "      <td>0.000172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>D014786</td>\n",
       "      <td>Vision Disorders</td>\n",
       "      <td>0.000161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>D014012</td>\n",
       "      <td>Tinnitus</td>\n",
       "      <td>0.000159</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier              name  weighted_DWPC\n",
       "0    D003128              Coma       0.000651\n",
       "1    D014474   Unconsciousness       0.000410\n",
       "2    D006935       Hypercapnia       0.000353\n",
       "3    D006261          Headache       0.000277\n",
       "4    D006985  Hyperventilation       0.000257\n",
       "5    D012021  Reflex, Abnormal       0.000256\n",
       "6    D015354       Vision, Low       0.000185\n",
       "7    D004172          Diplopia       0.000172\n",
       "8    D014786  Vision Disorders       0.000161\n",
       "9    D014012          Tinnitus       0.000159"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spoke_metapath = \"(n0:Gene)-[:ASSOCIATES_GaS]-(n1:Symptom)-[:PRESENTS_DpS]-(n2:Disease {identifier:'%s'})\"\n",
    "spoke_metapath = spoke_metapath % disease\n",
    "edges = [edge[2:-1] for edge in edge_pattern.findall(spoke_metapath)]\n",
    "degree_str = get_inner_dwpc_degree_str(edges)\n",
    "query = OVERALL_WEIGHTED_DWPC_QUERY % (gs_metapath, gs_filter, spoke_metapath, gs_metapath[:-1], gs_filter, spoke_metapath, degree_str, group_by_node, group_by_node, weight_prop, group_by_node, group_by_node)\n",
    "dwpc_df = composite.run_cypher(query)\n",
    "inner_nodes = list(dwpc_df.identifier.values[:3])\n",
    "plot_dwpc_paths(genelab, driver, spoke_metapath, group_by_node, gs_metapath, gs_filter, inner_nodes, max_show = 10)\n",
    "dwpc_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa15f0d6-1c4e-40f0-a724-7c05dc986ead",
   "metadata": {},
   "source": [
    "### Look at weighted DWPC for BiologicalProcess using upregulated genes in Space Flight v Control assays in the Eye\n",
    "![alt text](bp_dwpc_metagraph_eye.png \"BP DWPC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "04a380a4-802f-492e-81d9-6af58a47d6b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6ad07cfac3aa4ad8a1f58eba6d53cf26",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GraphWidget(layout=Layout(height='500px', width='100%'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>weighted_DWPC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0019835</td>\n",
       "      <td>cytolysis</td>\n",
       "      <td>3.962651e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0032200</td>\n",
       "      <td>telomere organization</td>\n",
       "      <td>2.245425e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0060333</td>\n",
       "      <td>type II interferon-mediated signaling pathway</td>\n",
       "      <td>2.112674e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0048877</td>\n",
       "      <td>homeostasis of number of retina cells</td>\n",
       "      <td>2.007509e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0051095</td>\n",
       "      <td>regulation of helicase activity</td>\n",
       "      <td>1.941231e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0006954</td>\n",
       "      <td>inflammatory response</td>\n",
       "      <td>1.627973e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0071357</td>\n",
       "      <td>cellular response to type I interferon</td>\n",
       "      <td>1.504975e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0000122</td>\n",
       "      <td>negative regulation of transcription by RNA po...</td>\n",
       "      <td>1.482725e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:1903706</td>\n",
       "      <td>regulation of hemopoiesis</td>\n",
       "      <td>1.445508e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0034340</td>\n",
       "      <td>response to type I interferon</td>\n",
       "      <td>1.363305e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   identifier                                               name  \\\n",
       "0  GO:0019835                                          cytolysis   \n",
       "1  GO:0032200                              telomere organization   \n",
       "2  GO:0060333      type II interferon-mediated signaling pathway   \n",
       "3  GO:0048877              homeostasis of number of retina cells   \n",
       "4  GO:0051095                    regulation of helicase activity   \n",
       "5  GO:0006954                              inflammatory response   \n",
       "6  GO:0071357             cellular response to type I interferon   \n",
       "7  GO:0000122  negative regulation of transcription by RNA po...   \n",
       "8  GO:1903706                          regulation of hemopoiesis   \n",
       "9  GO:0034340                      response to type I interferon   \n",
       "\n",
       "   weighted_DWPC  \n",
       "0   3.962651e-07  \n",
       "1   2.245425e-07  \n",
       "2   2.112674e-07  \n",
       "3   2.007509e-07  \n",
       "4   1.941231e-07  \n",
       "5   1.627973e-07  \n",
       "6   1.504975e-07  \n",
       "7   1.482725e-07  \n",
       "8   1.445508e-07  \n",
       "9   1.363305e-07  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eye_anatomies = ['UBERON:0000970', 'UBERON:0004548', 'UBERON:0000966']\n",
    "gs_metapath = \"(eye:Anatomy)-[]-(a:Assay)-[r:MEASURED_AmMG]->(:MGene)-[]->(n:Gene)\"\n",
    "gs_filter = \"\\nWHERE r.direction = 'upregulates' AND a.identifier IN %s AND eye.identifier IN %s\" % (sf_assays, eye_anatomies)\n",
    "\n",
    "spoke_metapath = \"(n0:Gene)-[:PARTICIPATES_GpBP]-(n1:BiologicalProcess)-[:PARTICIPATES_GpBP]-(n2:Gene)-[:ASSOCIATES_DaG]-(n3:Disease {identifier:'%s'})\"\n",
    "spoke_metapath = spoke_metapath % disease\n",
    "edges = [edge[2:-1] for edge in edge_pattern.findall(spoke_metapath)]\n",
    "degree_str = get_inner_dwpc_degree_str(edges)\n",
    "query = OVERALL_WEIGHTED_DWPC_QUERY % (gs_metapath, gs_filter, spoke_metapath, gs_metapath[:-1], gs_filter, spoke_metapath, degree_str, group_by_node, group_by_node, weight_prop, group_by_node, group_by_node)\n",
    "dwpc_df = composite.run_cypher(query)\n",
    "inner_nodes = list(dwpc_df.identifier.values[:8])\n",
    "plot_dwpc_paths(genelab, driver, spoke_metapath, group_by_node, gs_metapath, gs_filter, inner_nodes, max_show = 10)\n",
    "dwpc_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28a63ecd-7f85-448b-a01a-7fef0879749f",
   "metadata": {},
   "source": [
    "### Look at weighted DWPC for Symptoms using upregulated genes in Space Flight v Control assays in the Eye"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e1374572-d3f8-4d5f-9e98-522dde44f61c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e706e64f6f9748189b7abcb491daf78e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GraphWidget(layout=Layout(height='500px', width='100%'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>weighted_DWPC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D006261</td>\n",
       "      <td>Headache</td>\n",
       "      <td>0.000031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D004172</td>\n",
       "      <td>Diplopia</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D015354</td>\n",
       "      <td>Vision, Low</td>\n",
       "      <td>0.000008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D012021</td>\n",
       "      <td>Reflex, Abnormal</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D014786</td>\n",
       "      <td>Vision Disorders</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>D014012</td>\n",
       "      <td>Tinnitus</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>D001766</td>\n",
       "      <td>Blindness</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>D003128</td>\n",
       "      <td>Coma</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier              name  weighted_DWPC\n",
       "0    D006261          Headache       0.000031\n",
       "1    D004172          Diplopia       0.000010\n",
       "2    D015354       Vision, Low       0.000008\n",
       "3    D012021  Reflex, Abnormal       0.000007\n",
       "4    D014786  Vision Disorders       0.000007\n",
       "5    D014012          Tinnitus       0.000003\n",
       "6    D001766         Blindness       0.000002\n",
       "7    D003128              Coma       0.000002"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spoke_metapath = \"(n0:Gene)-[:ASSOCIATES_GaS]-(n1:Symptom)-[:PRESENTS_DpS]-(n2:Disease {identifier:'%s'})\"\n",
    "spoke_metapath = spoke_metapath % disease\n",
    "edges = [edge[2:-1] for edge in edge_pattern.findall(spoke_metapath)]\n",
    "degree_str = get_inner_dwpc_degree_str(edges)\n",
    "query = OVERALL_WEIGHTED_DWPC_QUERY % (gs_metapath, gs_filter, spoke_metapath, gs_metapath[:-1], gs_filter, spoke_metapath, degree_str, group_by_node, group_by_node, weight_prop, group_by_node, group_by_node)\n",
    "dwpc_df = composite.run_cypher(query)\n",
    "inner_nodes = list(dwpc_df.identifier.values[:3])\n",
    "plot_dwpc_paths(genelab, driver, spoke_metapath, group_by_node, gs_metapath, gs_filter, inner_nodes, max_show = 10)\n",
    "dwpc_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b0c806e-38b1-446b-9cd7-d2be275d9daf",
   "metadata": {},
   "source": [
    "# Make Projection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70900f6-2f17-4e1d-b186-003b2a60c4a1",
   "metadata": {},
   "source": [
    "![alt text](jupyter_example_images_projection_gds.png \"Projection\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "04ac5377-c5b8-4ceb-b182-7b3ff8cf6b73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nodeId</th>\n",
       "      <th>Node</th>\n",
       "      <th>Node_Name</th>\n",
       "      <th>Node_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2496884</td>\n",
       "      <td>9833</td>\n",
       "      <td>MELK</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001069</td>\n",
       "      <td>inchikey:JYGXADMDTFJGBT-ZSWZSQRESA-N</td>\n",
       "      <td>(9S,10R,11S,13S,14S,17R)-11,17-dihydroxy-17-(2...</td>\n",
       "      <td>Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2492538</td>\n",
       "      <td>3312</td>\n",
       "      <td>HSPA8</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2493458</td>\n",
       "      <td>4783</td>\n",
       "      <td>NFIL3</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2495116</td>\n",
       "      <td>7048</td>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    nodeId                                  Node  \\\n",
       "0  2496884                                  9833   \n",
       "1  1001069  inchikey:JYGXADMDTFJGBT-ZSWZSQRESA-N   \n",
       "2  2492538                                  3312   \n",
       "3  2493458                                  4783   \n",
       "4  2495116                                  7048   \n",
       "\n",
       "                                           Node_Name Node_Type  \n",
       "0                                               MELK      Gene  \n",
       "1  (9S,10R,11S,13S,14S,17R)-11,17-dihydroxy-17-(2...  Compound  \n",
       "2                                              HSPA8      Gene  \n",
       "3                                              NFIL3      Gene  \n",
       "4                                             TGFBR2      Gene  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from spokecloud import get_projection_info\n",
    "projection_name='genelab_proj'\n",
    "G=spoke.graph.get(projection_name)\n",
    "node_info_df = get_projection_info(spoke, projection_name)\n",
    "node_info_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20f4c9c4-7757-4285-a7db-52bb64383744",
   "metadata": {},
   "source": [
    "# Run PageRank\n",
    "##### The PageRank algorithm measures the importance of each node within the graph, based on the number incoming relationships and the importance of the corresponding source nodes. The underlying assumption roughly speaking is that a page is only as important as the pages that link to it. Here, we are creating a PageRank embedding for each comorbidity.\n",
    "![alt text](jupyter_example_images_pagerank.png \"PageRank\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b9823b7-accd-4dd3-9f94-c0a76907edca",
   "metadata": {},
   "source": [
    "#### Here we are using all upregulated genes in Space Flight v Control assays in the Eye as restart nodes and setting the dampingFactor = 0.33.\n",
    "#### Read more about how to configure PageRank [here](https://neo4j.com/docs/graph-data-science/current/algorithms/page-rank/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "27dcea39-3901-4c95-bdb2-68af70e0c46b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e77680def31c4399adfe9549f301ca9c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GraphWidget(layout=Layout(height='500px', width='100%'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nodeId</th>\n",
       "      <th>Node</th>\n",
       "      <th>Node_Name</th>\n",
       "      <th>Node_Type</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2496884</td>\n",
       "      <td>9833</td>\n",
       "      <td>MELK</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.769196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001069</td>\n",
       "      <td>inchikey:JYGXADMDTFJGBT-ZSWZSQRESA-N</td>\n",
       "      <td>(9S,10R,11S,13S,14S,17R)-11,17-dihydroxy-17-(2...</td>\n",
       "      <td>Compound</td>\n",
       "      <td>0.000779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2492538</td>\n",
       "      <td>3312</td>\n",
       "      <td>HSPA8</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.839952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2493458</td>\n",
       "      <td>4783</td>\n",
       "      <td>NFIL3</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.758840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2495116</td>\n",
       "      <td>7048</td>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.166235</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    nodeId                                  Node  \\\n",
       "0  2496884                                  9833   \n",
       "1  1001069  inchikey:JYGXADMDTFJGBT-ZSWZSQRESA-N   \n",
       "2  2492538                                  3312   \n",
       "3  2493458                                  4783   \n",
       "4  2495116                                  7048   \n",
       "\n",
       "                                           Node_Name Node_Type     score  \n",
       "0                                               MELK      Gene  0.769196  \n",
       "1  (9S,10R,11S,13S,14S,17R)-11,17-dihydroxy-17-(2...  Compound  0.000779  \n",
       "2                                              HSPA8      Gene  0.839952  \n",
       "3                                              NFIL3      Gene  0.758840  \n",
       "4                                             TGFBR2      Gene  0.166235  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from spokecloud import display_top_nodes, PAGERANK_WEIGHT_QUERY, PAGERANK_SINGLE_GENELIST_QUERY, run_multi_pagerank\n",
    "\n",
    "dampingFactor = 0.33\n",
    "filter_str = ''\n",
    "pagerank_df = composite.run_cypher(PAGERANK_SINGLE_GENELIST_QUERY%(gs_metapath, gs_filter, 'Gene', projection_name, dampingFactor))\n",
    "pagerank_df = node_info_df.merge(pagerank_df, on='nodeId')\n",
    "top_node_df = display_top_nodes(pagerank_df, driver, n_top=10, score_col='score', asc=False, db_name='spoke')\n",
    "pagerank_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ba131ffe-220f-4fb0-9d95-94ebf5147fa8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CALL {\n",
      "    USE compositenasa.glds\n",
      "    MATCH %s%s\n",
      "    RETURN COLLECT(DISTINCT n.identifier) AS ids\n",
      "}\n",
      "CALL {\n",
      "    USE compositenasa.gldsspoke\n",
      "    WITH ids\n",
      "    MATCH (g:%s) WHERE g.identifier IN ids\n",
      "    RETURN DISTINCT ID(g) AS overlap_genes\n",
      "}\n",
      "WITH COLLECT(overlap_genes) as sourceNodes\n",
      "CALL {\n",
      "    USE compositenasa.gldsspoke\n",
      "    WITH sourceNodes\n",
      "    CALL gds.pageRank.stream('%s', { sourceNodes:sourceNodes, maxIterations:40,dampingFactor:%s})\n",
      "    YIELD nodeId, score\n",
      "    RETURN nodeId, score \n",
      "}\n",
      "RETURN nodeId, score ORDER BY score DESC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(PAGERANK_SINGLE_GENELIST_QUERY)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f2f60f0-733b-41f0-96e0-24e94c95ec46",
   "metadata": {},
   "source": [
    "#### Here we are running PageRank for each upregulated genes in Space Flight v Control assays in the Eye. This results in an embedding vector for each gene which is then multiplied by the average log2fc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8fd44219-9f1f-4407-970e-776c177c579b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4c3aa7e5c6f0434eba779c9a9c92e373",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GraphWidget(layout=Layout(height='500px', width='100%'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pagerank_weight_df = composite.run_cypher(PAGERANK_WEIGHT_QUERY % (gs_metapath, gs_filter, weight_prop))\n",
    "multi_pagerank_df = run_multi_pagerank(spoke, G, node_info_df, pagerank_weight_df)\n",
    "top_node_df = display_top_nodes(multi_pagerank_df, driver, n_top=10, score_col='mean_pr', asc=False, db_name='spoke')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4ca78950-cbc1-4491-99ba-e2892230ba80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GLDS-173-ec2ef062472a03059fc9d18ffbeb6415',\n",
       " 'GLDS-21-a9111b1e09b475016847c828371b2c84',\n",
       " 'GLDS-21-150b1510550b43ee09451f2d8a0bd32c',\n",
       " 'GLDS-21-ec2ef062472a03059fc9d18ffbeb6415',\n",
       " 'GLDS-25-ec2ef062472a03059fc9d18ffbeb6415',\n",
       " 'GLDS-513-ec2ef062472a03059fc9d18ffbeb6415']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"MATCH (n:Assay) WHERE \"['Ground Control']\" in n.factors_1 RETURN DISTINCT n.identifier as n\"\"\"\n",
    "list(genelab.run_cypher(query).n.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "70047700-e100-4aa0-bb9b-be3b0e78b5a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([list([\"['Vivarium Control']\"]),\n",
       "       list([\"['Hindlimb Suspension and Reloading']\"]),\n",
       "       list([\"['Hindlimb Suspension']\"])], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"MATCH (n:Assay) WHERE \"['Ground Control']\" in n.factors_1 RETURN DISTINCT n.factors_2 as n\"\"\"\n",
    "genelab.run_cypher(query).n.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a2d0c3-de5e-4c07-bd93-3ca2626c1c4c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
